Adding new form factor and improving PR comments

Mirzamohammad22 opened the pull request:
On using the package i noticed some features could be slightly improved.

Issues:
1. Not being able to run both emulationFormFactors in the same call and see pr comments of both without replication of comments.
2. Pr comments not having information about which emulationFormFactor it ran on, anyone reviewing the pr would have to see the full report for it.

This Pr is meant to add these improvements:
1. Add a new 'both' emulationFormFactor for local lighthouse to run 'mobile' and 'desktop'
2. Add the 'mode' for pr comments so reviewers have more context to work with, without having to see the full report.

Mirzamohammad22 committed:
Adding new form factor,Updating pr comments

Mirzamohammad22 commented:
Hey @adamhenson .Awsome project, just some minor contributions based on experience on my end. Would love to have your suggestions on it.

adamhenson commented on package-lock.json:
```diff
@@ -3307,28 +3307,28 @@
       "dependencies": {
         "abbrev": {
           "version": "1.1.1",
-          "resolved": false,
```
Can we exclude changes in this file since no dependencies were changed?

Mirzamohammad22 replied:
Done

adamhenson commented on src/localLighthouse.js:
```diff
@@ -143,11 +163,7 @@ export default async ({
-    }
-
-    const lighthouseAuditResult = await localLighthouse({
+    let options = {
```
This should be a `const` (sorry the linting isn't better)

Mirzamohammad22 replied:
Yes, will do.

adamhenson commented on src/localLighthouse.js:
```diff
@@ -160,11 +176,20 @@ export default async ({
 
-    auditResults.push(lighthouseAuditResult);
-    index++;
+    if (options.emulatedFormFactor != 'both') {
```
Can we do a strict comparison for consistency (`!==`)? Sorry the linting doesn't catch this.

adamhenson replied:
Can we update the `README.md` file to add the `both` value?

```diff
- oneOf(['mobile', 'desktop'])
+ oneOf(['both', 'mobile', 'desktop'])
```

adamhenson replied:
Can we delete `lighthouseCaller` and instead just add the logging within the if / else blocks. And you could set the default option to `mobile` updating this line to `emulatedFormFactor = 'mobile'`:

https://github.com/foo-software/lighthouse-check/blob/78a9b0a60128d0764fc73dd77028773a49414fd9/src/lighthouseCheck.js#L20

Mirzamohammad22 replied:
Yes, i was waiting for your approval of source code before adding it to documentation.

Mirzamohammad22 replied:
> Can we delete `lighthouseCaller` and instead just add the logging within the if / else blocks. And you could set the default option to `mobile` updating this line to `emulatedFormFactor = 'mobile'`:
> 
> https://github.com/foo-software/lighthouse-check/blob/78a9b0a60128d0764fc73dd77028773a49414fd9/src/lighthouseCheck.js#L20
I added lighthouse caller for a more DRY approach (IMO) instead of having multiple prints, i will update default emulatedFormFactor. Let me know if i should still remove the function

adamhenson commented on src/localLighthouse.js:
```diff
@@ -5,6 +5,25 @@ import lighthouseDefaultConfig, { throttling } from './lighthouseConfig';
 import options from './lighthouseOptions';
 import writeResults from './helpers/writeResults';
 import { NAME } from './constants';
+import { errorMonitor } from 'stream';
```
I don't think this is used (sorry the linter doesn't catch this).

adamhenson commented on src/postPrComment.js:
```diff
@@ -31,10 +31,12 @@ export default async ({
       markdown += `\n\n${result.url}`;
 
+      // the emulatedformfactor
+      markdown += `\n\n Mode: **${result.emulatedFormFactor}**`;
```
I don't think this will work because `emulatedFormFactor` won't be part of `result` (unless I'm mistaken). We should have a test for comments, so I opened [issue #33](https://github.com/foo-software/lighthouse-check/issues/33).

Mirzamohammad22 replied:
i have added it in the localLighthouse function `return` object  and tested it, looks fine.
<img width="1026" alt="Screen Shot 2020-06-24 at 6 14 27 PM" src="https://user-images.githubusercontent.com/38260994/85573241-c5293700-b646-11ea-8241-76677878279d.png">

adamhenson replied:
Super minor - can we put the link below "Mode: mobile" @Mirzamohammad22?

Mirzamohammad22 replied:
Sure

adamhenson commented:
Thanks for opening this @Mirzamohammad22. I think this could be a beneficial feature. I left a few comments. I'll work on [issue #33](https://github.com/foo-software/lighthouse-check/issues/33) so we can test it.

Mirzamohammad22 committed:
Updating README , localLighthouse

adamhenson commented on src/localLighthouse.js:
```diff
@@ -108,6 +127,7 @@ export const localLighthouse = async ({
     url,
     localReport,
     report,
+    emulatedFormFactor,
```
👍

Mirzamohammad22 committed:
Updating PrComment,lighthouseCheck,localLighthouse

Mirzamohammad22 committed:
Adding old package-lock

Mirzamohammad22 commented:
@adamhenson Would love to have these improvements in your lighthouse check action.

adamhenson commented on src/localLighthouse.js:
```diff
@@ -6,6 +6,20 @@ import options from './lighthouseOptions';
+    const lighthouseAuditResult = await localLighthouse(options);
+    return lighthouseAuditResult;
+  } catch (err) {
+    console.log(err);
```
We don't want to catch the error here because an error here means something is in fact wrong and it should stop the process. We can also leave the other `console.log` where it was before (outside of this function...but update the message). And then with those changes we can just delete this function and use `localLighthouse` directly.

adamhenson commented on src/localLighthouse.js:
```diff
@@ -143,11 +158,7 @@ export default async ({
   let index = 1;
 
   for (const url of urls) {
-    if (verbose) {
```
We can now keep this block as is, but just update the message.

adamhenson commented on README.md:
```diff
@@ -503,7 +503,7 @@ Below are options for the exported `lighthouseCheck` function or `lighthouse-che
     <td><code>emulatedFormFactor</code></td>
     <td>Lighthouse setting only used for local audits. See <a href="src/lighthouseConfig.js">src/lighthouseConfig.js</a> comments for details.</td>
-    <td><code>oneOf(['mobile', 'desktop']</code></td>
+    <td><code>oneOf(['mobile', 'desktop','both']</code></td>
```
I know this was an issue before, but now that we're updating this can we add the closing `)`... like so?
```
<td><code>oneOf(['mobile', 'desktop','both'])</code></td>
```

adamhenson commented:
Thanks for making these changes @Mirzamohammad22 - I just added a couple last ones. I think we should remove `lighthouseCaller` because we don't need the console log inside of it. Since we added the default `emulatedFormFactor = 'mobile'`, the console log outside of that function works as is 👍. If I'm not making sense - I could create a PR into your branch maybe.

adamhenson commented:
Yep, once we get this merged, I'll publish a new version and bump the GitHub Action to consume it.

Mirzamohammad22 commented:
> Thanks for making these changes @Mirzamohammad22 - I just added a couple last ones. I think we should remove `lighthouseCaller` because we don't need the console log inside of it. Since we added the default `emulatedFormFactor = 'mobile'`, the console log outside of that function works as is 👍. If I'm not making sense - I could create a PR into your branch maybe.
I'll make the changes, no issues.

Mirzamohammad22 committed:
Updating localLighthouse, README

Mirzamohammad22 committed:
Removing lighthouseCaller

adamhenson commented on src/localLighthouse.js:
```diff
@@ -160,11 +157,38 @@ export default async ({
-    auditResults.push(lighthouseAuditResult);
-    index++;
+    if (options.emulatedFormFactor !== 'both') {
+      if (verbose) {
```
~Sorry if I wasn't clear @Mirzamohammad22. My point is that we don't need to have this block repeated now. We can leave it where it was.~

EDITED: My mistake, I see why we need it, because `options` is changed.

Mirzamohammad22 replied:
No issues , let me know  if the `If case` is fine or should i revert to lighthouseCaller

adamhenson commented:
@Mirzamohammad22 thanks again for your work on this. I'm merging this into another branch I created, so I can make some minor changes. I'll let you know when I've merged and published it (should be EOD).

Mirzamohammad22 commented:
> @Mirzamohammad22 thanks again for your work on this. I'm merging this into another branch I created, so I can make some minor changes. I'll let you know when I've merged and published it (should be EOD).

Awesome.Thanks for the approval , looking forward to having this feature in GitHub actions!

adamhenson commented:
@Mirzamohammad22 this has been released added to the GitHub Action in [v2.0.1](https://github.com/foo-software/lighthouse-check-action/releases/tag/v2.0.1) 🎈 - thanks again!

adamhenson commented:
Note - I updated `emulatedFormFactor` to accept `all` vs `both`.

Mirzamohammad22 commented:
> @Mirzamohammad22 this has been released added to the GitHub Action in [v2.0.1](https://github.com/foo-software/lighthouse-check-action/releases/tag/v2.0.1) 🎈 - thanks again!

Thank you!
